caput/task/io.py

```diff
@@ -0,0 +1,1018 @@
+
+from .. import config, fileformats, memh5, pipeline, truncate
+from ._core import MPILoggedTask, SingleTask
+
+
+class ConfigError(Exception):
+    """Raised due to an error with a task configuration."""
+
+    pass
+
+
+def list_of_filelists(files: list[str] | list[list[str]]) -> list[list[str]]:
+    """Take in a list of lists/glob patterns of filenames.
+
+    Parameters
+    ----------
+    files
+        A path or glob pattern (e.g. /my/data/*.h5) or a list of those (or a list of
+        lists of those).
+
+    Raises
+    ------
+    ConfigError
+        If files has the wrong format or refers to a file that doesn't exist.
+
+    Returns
+    -------
+    The input file list list. Any glob patterns will be flattened to file path string
+    lists.
+    """
+    import glob
+
+    f2 = []
+
+    for filelist in files:
+        if isinstance(filelist, str):
+            if "*" not in filelist and not os.path.isfile(filelist):
+                raise ConfigError(f"File not found: {filelist!s}")
+            filelist = glob.glob(filelist)
+        elif isinstance(filelist, list):
+            for i in range(len(filelist)):
+                filelist[i] = list_or_glob(filelist[i])
+        else:
+            raise ConfigError("Must be list or glob pattern.")
+        f2 = f2 + filelist
+
+    return f2
+
+
+def list_or_glob(files: str | list[str]) -> list[str]:
+    """Take in a list of lists/glob patterns of filenames.
+
+    Parameters
+    ----------
+    files
+        A path or glob pattern (e.g. /my/data/*.h5) or a list of those
+
+    Returns
+    -------
+    The input file list. Any glob patterns will be flattened to file path string lists.
+
+    Raises
+    ------
+    ConfigError
+        If files has the wrong type or if it refers to a file that doesn't exist.
+    """
+    import glob
+
+    # If the input was a list, process each element and return as a single flat list
+    if isinstance(files, list):
+        parsed_files = []
+        for f in files:
+            parsed_files = parsed_files + list_or_glob(f)
+        return parsed_files
+
+    # If it's a glob we need to expand the glob and then call again
+    if isinstance(files, str) and any(c in files for c in "*?["):
+        return list_or_glob(sorted(glob.glob(files)))
+
+    # We presume a string is an actual path...
+    if isinstance(files, str):
+        # Check that it exists and is a file (or dir if zarr format)
+        if files.endswith(".zarr"):
+            if not os.path.isdir(files):
+                raise ConfigError(
+                    f"Expected a zarr directory store, but directory not found: {files}"
+                )
+        else:
+            if not os.path.isfile(files):
+                raise ConfigError(f"File not found: {files!s}")
+
+        return [files]
+
+    raise ConfigError(
+        f"Argument must be list, glob pattern, or file path, got {files!r}"
+    )
+
+
+def list_of_filegroups(groups: list[dict] | dict) -> list[dict]:
```

Comment:
As with `list_of_filelists`, is this used anywhere other than `draco.core.io`?  Is there a reason not to just leave it there?

Reply:
Ah, It's used by `ch_pipeline`.

Reply:
Yeah, these could maybe go somewhere else but I'm not really sure what the best place for them is.